### PR TITLE
Add sphinx doc for python bindings

### PR DESCRIPTION
**Pull request by tfcollins**
Users have been asking more and more about these bindings and the "magic" calls we sometimes do in pyadio-iio. This doc should help.

Open question is really around hosting. Ideally this would go on github pages along side doxygen (somehow), but we can go to RTD in the short term.

Just don't want to forget about this feature.

Signed-off-by: Travis F. Collins <travis.collins@analog.com>

### CODE REVIEW

File: bindings/python/doc/conf.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import sys
+sys.path.insert(0, os.path.abspath('../.'))
+
+import sphinx_rtd_theme
+
+# -- Project information -----------------------------------------------------
+
+project = 'libiio'
+copyright = '2020, Analog Devices, Inc.'
+author = 'Paul Cercueil <paul.cercueil@analog.com>'
```
Comment (commodo) [inline bindings/python/doc/conf.py]: curios: do we want to keep using Paul's non-working email?
or is it working now;
i did send some kernel patches using emails, and I got a lot of reply-backs that it wasn't working;

**Comment by commodo** [comment]
rest looks good for now;
any place where we can see the output?
or is that later?

File: bindings/python/doc/conf.py (hunk 1 above)
Comment (tfcollins) [reply bindings/python/doc/conf.py]: Ill remove the email

**Comment by tfcollins** [comment]
You will have to build it yourself at the moment, thats more of an open question right now since it will require changing how we manage the doxygen (if we want to put them in the same place), and I would suggest we automatically update doxygen on releases. Probably having a version of doc for each release and able to go back would be useful too.

File: bindings/python/doc/conf.py (hunk 1 above)
Comment (rgetz) [reply bindings/python/doc/conf.py]: Thats a good point - does it need to get updated in all the fies?

**Comment by rgetz** [comment]
yeah, I have been thinking of that for a bit... I can have a look about hosting things in the ghpages, in a subdirectory per release for now...

**Comment by rgetz** [comment]
What's needed to build?
```
sudo pip install --user sphinx
```

looks like this is missing the CMake hooks to build the makefile, something like some of the parts in:
https://github.com/Kitware/CMake/blob/master/Utilities/Sphinx/CMakeLists.txt#L127

It's easier if people can just use an existing CMake flag (WITH_DOC && PYTHON_BINDINGS), and if things find the SPHINX_EXECUTABLE - it just builds...

Thanks
-Robin

**Comment by tfcollins** [comment]
Yeah, just build the doc if it exists

**Comment by tfcollins** [comment]
https://github.com/koumoul-dev/gh-pages-multi

**Comment by rgetz** [comment]
> Yeah, just build the doc if it exists

Really? I tried:

cmake WITH_DOC=1 PYTHON_BINDINGS=1 && make

and I have sphinx-build on path, and I didn't get any doc...

**Comment by tfcollins** [comment]
O sorry, I meant that's what the behavior should be :). Go into the python/doc folder and type `make html`

**Comment by rgetz** [comment]
```
rgetz@brain:~/github/libiio/bindings/python/doc$ make html
make: *** No rule to make target 'html'.  Stop.
```

to get the Makefile I needed to do something like:

```
rgetz@brain:~/github/libiio/bindings/python/doc$ rm ../../../build/html/* -rf
rgetz@brain:~/github/libiio/bindings/python/doc$ sphinx-apidoc -o ../../../build/html/ ./
Creating file ../../../build/html/conf.rst.
Creating file ../../../build/html/modules.rst.
rgetz@brain:~/github/libiio/bindings/python/doc$ sphinx-apidoc -o ../../../build/html/ ./ -F
File ../../../build/html/conf.rst already exists, skipping.
Creating file ../../../build/html/conf.py.
Creating file ../../../build/html/index.rst.
Creating file ../../../build/html/Makefile.
Creating file ../../../build/html/make.bat.
rgetz@brain:~/github/libiio/bindings/python/doc$ cd ../../../build/html/
rgetz@brain:~/github/libiio/build/html$ make html
Running Sphinx v1.8.5
loading translations [en]... done
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 3 source files that are out of date
updating environment: 3 added, 0 changed, 0 removed
reading sources... [100%] modules                                                                                          
WARNING: autodoc: failed to import module u'conf'; the following exception was raised:
No module named conf
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/rgetz/github/libiio/build/html/modules.rst: WARNING: document isn't included in any toctree
done
preparing documents... done
writing output... [100%] modules                                                                                           
generating indices... genindex
writing additional pages... search
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 2 warnings.

The HTML pages are in _build/html.
```

**Comment by rgetz** [comment]
After a 
```
pip install sphinx-rtd-theme
```

I get:
```
rgetz@brain:~/github/libiio/bindings/python/doc$ make html
Running Sphinx v1.8.5
making output directory...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 7 source files that are out of date
updating environment: 7 added, 0 changed, 0 removed
//home/rgetz/.local/lib/python2.7/site-packages/sphinx/util/nodes.py:94: FutureWarning:                                   
   The iterable returned by Node.traverse()
   will become an iterator instead of a list in Docutils > 0.16.
  for classifier in reversed(node.parent.traverse(nodes.classifier)):
reading sources... [100%] trigger                                                                                         

Sphinx error:
master file /home/rgetz/github/libiio/bindings/python/doc/contents.rst not found
make: *** [Makefile:20: html] Error 2
```

I needed to do:
```
rgetz@brain:~/github/libiio/bindings/python/doc$ git diff
diff --git a/bindings/python/doc/conf.py b/bindings/python/doc/conf.py
index 14e7574..cca32eb 100644
--- a/bindings/python/doc/conf.py
+++ b/bindings/python/doc/conf.py
@@ -54,3 +54,5 @@ html_theme = "sphinx_rtd_theme"
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ['_static']
+
+master_doc = 'index'
```

to get rid of it, it looks like https://www.sphinx-doc.org/en/master/usage/configuration.html?highlight=master_doc says _Changed in version 2.0: The default is changed to 'index' from 'contents'._

but then yeah, it seems to work from there...

**Comment by tfcollins** [comment]
Is that on windows or linux?

**Comment by rgetz** [comment]
On linux - Debian stable

**Comment by tfcollins** [comment]
Built fine on raspbian.

```
pi@raspberrypi:~ $ pip3 show sphinx
Name: Sphinx
Version: 2.4.4
```

File: bindings/python/doc/conf.py (hunk 1 above)
Comment (tfcollins) [reply bindings/python/doc/conf.py:23]: @rgetz you mean in the non-python files too?

File: bindings/python/doc/conf.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import sphinx_rtd_theme
+
+# -- Project information -----------------------------------------------------
+
+project = 'libiio-py'
+copyright = '2020, Analog Devices, Inc.'
+author = 'Paul Cercueil <paul.cercueil@analog.com>'
+
+# The full version, including alpha/beta/rc tags
+release = '0.19.g8af9e04'
```
Comment (tfcollins) [inline bindings/python/doc/conf.py:26]: Is this right?

File: bindings/python/doc/conf.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import sys
+sys.path.insert(0, os.path.abspath('../.'))
+
+import sphinx_rtd_theme
+
+# -- Project information -----------------------------------------------------
+
+project = 'libiio-py'
+copyright = '2020, Analog Devices, Inc.'
+author = 'Paul Cercueil <paul.cercueil@analog.com>'
```
Comment (tfcollins) [inline bindings/python/doc/conf.py:23]: Can we delete this file now?

File: bindings/python/CMakeLists.txt [hunk 4]
```diff
@@ -29,3 +29,26 @@ if (Python_Interpreter_FOUND)
...
+
+if(WITH_DOC)
+	find_program(SPHINX_EXECUTABLE
+		NAMES sphinx-build
+		DOC "Sphinx Documentation Builder (sphinx-doc.org)"
+	)
+	if (NOT SPHINX_EXECUTABLE)
+		message(FATAL_ERROR "Can not build python doc without sphinx-build")
+	endif()
+	message (STATUS "building with Python Doc (Sphinx)")
```
Comment (tfcollins) [inline bindings/python/CMakeLists.txt:41]: Nit pick :( capitalize b in building

File: bindings/python/doc/conf.py (hunk 3 above)
Comment (rgetz) [reply bindings/python/doc/conf.py:23]: I thought I did...

File: bindings/python/CMakeLists.txt (hunk 4 above)
Comment (rgetz) [reply bindings/python/CMakeLists.txt]: fixed.

File: bindings/python/doc/conf.py (hunk 3 above)
Comment (rgetz) [reply bindings/python/doc/conf.py]: Fixed

**Comment by rgetz** [comment]
@commodo : it's building on the CI now - next release it will be on gh-pages (assuming that most of this is acceptable), I can temporarily fork this, and push gh-pages up if you want to have a look...

File: CI/travis/make_linux [hunk 5]
```diff
@@ -37,6 +24,15 @@ handle_default() {
 		echo "### Deleting install_manifest.txt (created as root)"
 		sudo rm ./install_manifest.txt
 	fi
+	if command_exists sphinx-build ; then
```
Comment (commodo) [inline CI/travis/make_linux:27]: do we need to build docs on all deb builds or should we split the build into a `handle_doc_build()`?
the `handle_default()` rule looks complicated and gets run for all Ubuntu build-runs;
if we can simplify it a bit, would be neat;

Comment (rgetz) [reply CI/travis/make_linux:27]: it depends on how many of the deb/rpms/tarballs we want with the doc, or not.

If we decide that the docs are all on-line - and they don't exist at all on the deb/rpm - that's an option.

**Pull request merged**